### DISCUSSION TITLE
`AddHeadersToRequest` with Claims[sub] and Claims[roles]

### DISCUSSION THREAD

**Discussion by jnormen** [ThreeMammals/Ocelot#1939] [Show and tell]
## Expected Behavior / New Feature
When pass this:
```json
"AddHeadersToRequest": {
   "tenantId": "Claims[tenantId] > value",
    "caller": "Claims[sub] > value",
    "roles": "Claims[roles] > value"
}
```
with this token:
```
"sub": "8723c40a-0b57-4447-b1a5-d3c15f7ca9ea",
"typ": "Bearer",
"roles": ["Admin","User"]
```
Expect to have 3 header values. 

## Actual Behavior / Motivation for New Feature
```
  "tenantId": "Claims[tenantId] > value",
```
This works no problem. If I type `typ` it also works. if I type other claims from token it all works. 

But when I use `sub` it says: `Cannot find a claim for key: sub errors found in ResponderMiddleware.`

If `roles`, the same: `Cannot find a claim for key: roles errors found in ResponderMiddleware.`

This they are all there in the token. And I can use most of the other claim names.

Is `sub` and `roles` somewhat reserved or ignored? 

## Steps to Reproduce the Problem

  1. add: 
  ```json
  "AddHeadersToRequest": {
    "tenantId": "Claims[tenantId] > value",
    "caller": "Claims[sub] > value"
  }
  ```
  2.
  ```json
  "AuthenticationOptions": {
    "AuthenticationProviderKey": "myKey",
    "AllowedScopes": []
  }
  ```
  3. call API with valid bearer token.

**Comment by jnormen** [comment]
I see now that JwtBearer settings in ASP .Net with the Jwt Schema change sub and roles to namespaces like this: {http://schemas.xmlsoap.org/ws/2005/05/identity/claims/nameidentifier: 8723c40a-0b57-4447-b1a5-d3c15f7ca9ea}

so my sub I pass in is transformed to "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/nameidentifier"

To fix this I need to do:
var claimValues = claims.Where(c => c.Type == ClaimTypes.NameIdentifier).Select(c => c.Value).ToArray();

in the ClaimsParser if key is sub.
Or use:
"caller" : "Claims[http://schemas.xmlsoap.org/ws/2005/05/identity/claims/nameidentifier]

**Reply by raman-m** [reply]
@jnormen 
Do you still see an issue or not? I don't!
We can discuss it more, if you don't like your hacky fix... :wink:

**Reply by jnormen** [reply]
This is long time ago and I do no use Ocelot any more. 
2024 now  :) 2019 was long time ago in software world :)

**Reply by raman-m** [reply]
I know, long time ago... 
But it was your feature request, right? Or it uncovers a bug in Ocelot or in ASP.NET framework...
I guess the issue is still actual...

**Comment by EasyMilos** [comment]
Or you use 
`JwtSecurityTokenHandler.DefaultInboundClaimTypeMap.Clear();`
in Startup and claims are back normal.

**Reply by raman-m** [reply]
Well... Could we create complete list of reserved claim names somehow?
And how to detect wrong claim name during run-time?

**Comment by Navaladi2019** [comment]
if the claim name is `nameid`, it is not getting forwarded. On other names it is working fine

**Reply by raman-m** [reply]
Please, prove that fact by a link to ASP.NET official repository!
